<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="fishtree">
<title>A simple comparative analysis with the Fish Tree of Life • fishtree</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A simple comparative analysis with the Fish Tree of Life">
<meta property="og:description" content="fishtree">
<meta name="twitter:card" content="summary">
<meta name="twitter:creator" content="@chang_jon">
<meta name="twitter:site" content="@chang_jon">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">fishtree</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.4.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/community-analysis.html">Phylogenetic community assembly with the Fish Tree of Life</a>
    <a class="dropdown-item" href="../articles/comparative-analysis.html">A simple comparative analysis with the Fish Tree of Life</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jonchang/fishtree/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>A simple comparative analysis with the Fish Tree of Life</h1>
                        <h4 data-toc-skip class="author">Jonathan
Chang</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jonchang/fishtree/blob/HEAD/vignettes/comparative-analysis.Rmd" class="external-link"><code>vignettes/comparative-analysis.Rmd</code></a></small>
      <div class="d-none name"><code>comparative-analysis.Rmd</code></div>
    </div>

    
    
<p>To demonstrate the functionality of the <code>fishtree</code> package
and how it integrates well with the rest of the R phylogenetics
ecosystem, this vignette will walk you through a simple comparative
analysis.</p>
<div class="section level2">
<h2 id="getting-and-cleaning-data">Getting and cleaning data<a class="anchor" aria-label="anchor" href="#getting-and-cleaning-data"></a>
</h2>
<p>A common hypothesis tested in comparative methods is whether habitat
shifts drive rates of diversification in various groups. For example, <a href="https://doi.org/10.1111/jeb.12112" class="external-link">Santini et al. (2013)</a>
tested, among other things, whether reef-associated pufferfishes enjoyed
faster rates of speciation compared to their non-reef relatives.</p>
<p>First, load the <code>fishtree</code> package and download the subset
of the Fish Tree of Life corresponding to the taxon of interest. To make
things more interesting we’ll work on the entire order (<a href="https://fishtreeoflife.org/taxonomy/order/Tetraodontiformes/" class="external-link">Tetraodontiformes</a>)
rather than the family in the 2013 study.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fishtreeoflife.org/" class="external-link">fishtree</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fishtree_phylogeny.html">fishtree_phylogeny</a></span><span class="op">(</span>rank <span class="op">=</span> <span class="st">"Tetraodontiformes"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tree</span>, show.tip.label <span class="op">=</span> <span class="cn">FALSE</span>, no.margin <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="comparative-analysis_files/figure-html/get_phylo-1.png" width="672"></p>
<p>Next, we need to get habitat data and associate it with our
phylogeny. The <code>rfishbase</code> package (<a href="https://doi.org/10.1111/j.1095-8649.2012.03464.x" class="external-link">Boettiger et
al. 2012</a>) has a variety of convenient functions to access data
recorded by the Fishbase editors. Load the <code>rfishbase</code>
package and retrieve the relevant data in the <code>DemersPelag</code>
field, which identifies whether a species is reef-associated or not,
among other things.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/rfishbase/" class="external-link">rfishbase</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">tips</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">" "</span>, <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fb_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/rfishbase/reference/species.html" class="external-link">species</a></span><span class="op">(</span>species_list <span class="op">=</span> <span class="va">tips</span>, fields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Species"</span>, <span class="st">"DemersPelag"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(SpecCode)`</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fb_results</span> <span class="op">&lt;-</span> <span class="va">fb_results</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">fb_results</span><span class="op">$</span><span class="va">DemersPelag</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fb_results</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   Species                       DemersPelag    </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Abalistes stellatus           demersal       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Acanthaluteres spilomelanurus demersal       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Acanthaluteres vittiger       demersal       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Acanthostracion quadricornis  reef-associated</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Aluterus heudelotii           demersal       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Aluterus monoceros            reef-associated</span></span></code></pre></div>
<p>Note that we had to replace the underscores in the tip labels with
spaces. This is a common source of errors, so if your analyses don’t
seem to work correctly always check whether the functions you’re using
expect underscores or spaces.</p>
<p>There’s a lot of data in the <code>DemersPelag</code> field, but we
only want to know if the species is reef-associated or not.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>tip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">" "</span>, <span class="st">"_"</span>, <span class="va">fb_results</span><span class="op">$</span><span class="va">Species</span><span class="op">)</span>,</span>
<span>                   is_reef <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">fb_results</span><span class="op">$</span><span class="va">DemersPelag</span> <span class="op">==</span> <span class="st">"reef-associated"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">reef</span><span class="op">)</span></span>
<span><span class="co">#&gt;                             tip is_reef</span></span>
<span><span class="co">#&gt; 1           Abalistes_stellatus       0</span></span>
<span><span class="co">#&gt; 2 Acanthaluteres_spilomelanurus       0</span></span>
<span><span class="co">#&gt; 3       Acanthaluteres_vittiger       0</span></span>
<span><span class="co">#&gt; 4  Acanthostracion_quadricornis       1</span></span>
<span><span class="co">#&gt; 5           Aluterus_heudelotii       0</span></span>
<span><span class="co">#&gt; 6            Aluterus_monoceros       1</span></span></code></pre></div>
<p>We’ve also converted the tip labels back to underscores, since we
need to ensure that the tip labels on our phylogeny match the labels on
our trait data. The <code>geiger</code> package (<a href="https://doi.org/10.1093/bioinformatics/btu181" class="external-link">Pennell et
al. 2014</a>) provides a convenient function that will perform this
check. The <code>name.check</code> function expects row names on our
data object, so we will do that as well.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">geiger</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ape</span></span>
<span><span class="co">#&gt; Loading required package: phytools</span></span>
<span><span class="co">#&gt; Loading required package: maps</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">reef</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">reef</span><span class="op">$</span><span class="va">tip</span></span>
<span><span class="va">nc</span> <span class="op">&lt;-</span> <span class="fu">geiger</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/geiger/man/name.check.html" class="external-link">name.check</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">reef</span><span class="op">)</span></span>
<span><span class="va">nc</span></span>
<span><span class="co">#&gt; $tree_not_data</span></span>
<span><span class="co">#&gt;  [1] "Abalistes_stellaris"                   </span></span>
<span><span class="co">#&gt;  [2] "Acanthostracion_polygonius"            </span></span>
<span><span class="co">#&gt;  [3] "Chelonodon_patoca"                     </span></span>
<span><span class="co">#&gt;  [4] "Chelonodon_pleurospilus"               </span></span>
<span><span class="co">#&gt;  [5] "Chilomycterus_spinosus_spinosus"       </span></span>
<span><span class="co">#&gt;  [6] "Lagocephalus_lagocephalus_lagocephalus"</span></span>
<span><span class="co">#&gt;  [7] "Monotrete_cochinchinensis"             </span></span>
<span><span class="co">#&gt;  [8] "Monotrete_leiurus"                     </span></span>
<span><span class="co">#&gt;  [9] "Ostracion_cubicus"                     </span></span>
<span><span class="co">#&gt; [10] "Ostracion_immaculatus"                 </span></span>
<span><span class="co">#&gt; [11] "Ostracion_solorensis"                  </span></span>
<span><span class="co">#&gt; [12] "Paramonacanthus_filicauda"             </span></span>
<span><span class="co">#&gt; [13] "Rhinesomus_triqueter"                  </span></span>
<span><span class="co">#&gt; [14] "Sphoeroides_cheesemanii"               </span></span>
<span><span class="co">#&gt; [15] "Stephanolepis_auratus"                 </span></span>
<span><span class="co">#&gt; [16] "Stephanolepis_hispidus"                </span></span>
<span><span class="co">#&gt; [17] "Takifugu_fasciatus"                    </span></span>
<span><span class="co">#&gt; [18] "Tetraodon_abei"                        </span></span>
<span><span class="co">#&gt; [19] "Tetraodon_baileyi"                     </span></span>
<span><span class="co">#&gt; [20] "Tetraodon_biocellatus"                 </span></span>
<span><span class="co">#&gt; [21] "Tetraodon_cambodgiensis"               </span></span>
<span><span class="co">#&gt; [22] "Tetraodon_cutcutia"                    </span></span>
<span><span class="co">#&gt; [23] "Tetraodon_erythrotaenia"               </span></span>
<span><span class="co">#&gt; [24] "Tetraodon_fluviatilis"                 </span></span>
<span><span class="co">#&gt; [25] "Tetraodon_nigroviridis"                </span></span>
<span><span class="co">#&gt; [26] "Tetraodon_palembangensis"              </span></span>
<span><span class="co">#&gt; [27] "Tetraodon_suvattii"                    </span></span>
<span><span class="co">#&gt; [28] "Tetraodon_turgidus"                    </span></span>
<span><span class="co">#&gt; [29] "Tetrosomus_fornasini"                  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $data_not_tree</span></span>
<span><span class="co">#&gt; character(0)</span></span></code></pre></div>
<p>We’ve identified a mismatch between the tree and the data. We’ll
exclude the tips lacking trait data using <code>drop.tip</code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/drop.tip.html" class="external-link">drop.tip</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">nc</span><span class="op">$</span><span class="va">tree_not_data</span><span class="op">)</span></span></code></pre></div>
<p>If we also had data that was not in the tree, we could exclude that
using the following command, but it isn’t necessary in this case:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reef</span> <span class="op">&lt;-</span> <span class="va">reef</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">reef</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">nc</span><span class="op">$</span><span class="va">data_not_tree</span>, <span class="op">]</span></span></code></pre></div>
<p>Confirm that we have the same number of observations in the tree and
the data:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/summary.phylo.html" class="external-link">Ntip</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">reef</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plotting-diversification-rates">Plotting diversification rates<a class="anchor" aria-label="anchor" href="#plotting-diversification-rates"></a>
</h2>
<p>There are several other data sources available in the
<code>fishtree</code> package, including speciation rates computed via
the DR method (<a href="https://doi.org/10.1038/nature11631" class="external-link">Jetz et
al. 2012</a>). Retrieve speciation rate data:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fishtree_tip_rates.html">fishtree_tip_rates</a></span><span class="op">(</span>rank <span class="op">=</span> <span class="st">"Tetraodontiformes"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rates</span><span class="op">)</span></span>
<span><span class="co">#&gt;                           species  lambda.tv      mu.tv  lambda.tc      mu.tc</span></span>
<span><span class="co">#&gt; 2             Abalistes stellaris 0.08859469 0.01282721 0.09181900 0.02239813</span></span>
<span><span class="co">#&gt; 3             Abalistes stellatus 0.08859469 0.01282721 0.09181900 0.02239813</span></span>
<span><span class="co">#&gt; 34  Acanthaluteres spilomelanurus 0.10180976 0.01614045 0.16054060 0.06301037</span></span>
<span><span class="co">#&gt; 35        Acanthaluteres vittiger 0.10180976 0.01614045 0.16054060 0.06301037</span></span>
<span><span class="co">#&gt; 131    Acanthostracion polygonius 0.08549937 0.01178475 0.07697343 0.01359198</span></span>
<span><span class="co">#&gt; 132  Acanthostracion quadricornis 0.08549937 0.01178475 0.07697343 0.01359198</span></span>
<span><span class="co">#&gt;             dr</span></span>
<span><span class="co">#&gt; 2   0.11678851</span></span>
<span><span class="co">#&gt; 3   0.11909205</span></span>
<span><span class="co">#&gt; 34  0.28175984</span></span>
<span><span class="co">#&gt; 35  0.28406568</span></span>
<span><span class="co">#&gt; 131 0.07606301</span></span>
<span><span class="co">#&gt; 132 0.07213440</span></span></code></pre></div>
<p>We’re interested in just the <code>dr</code> column, so extract that
and convert spaces to underscores again. Then merge the habitat data
with the speciation rate data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>tip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">" "</span>, <span class="st">"_"</span>, <span class="va">rates</span><span class="op">$</span><span class="va">species</span><span class="op">)</span>, dr <span class="op">=</span> <span class="va">rates</span><span class="op">$</span><span class="va">dr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">rates</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">rates</span><span class="op">$</span><span class="va">tip</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">reef</span>, <span class="va">rates</span><span class="op">)</span></span></code></pre></div>
<p>As a quick check our data, let’s plot histograms of the DR rate of
reef and non-reef species:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">breaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">merged</span><span class="op">$</span><span class="va">dr</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">merged</span><span class="op">$</span><span class="va">dr</span><span class="op">)</span>, length.out <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">merged</span>, <span class="va">is_reef</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">$</span><span class="va">dr</span>, col <span class="op">=</span> <span class="st">"orange"</span>, density <span class="op">=</span> <span class="fl">20</span>, angle <span class="op">=</span> <span class="fl">135</span>,</span>
<span>     breaks <span class="op">=</span> <span class="va">breaks</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">merged</span>, <span class="va">is_reef</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">$</span><span class="va">dr</span>, col <span class="op">=</span> <span class="st">"purple"</span>, density <span class="op">=</span> <span class="fl">20</span>, angle <span class="op">=</span> <span class="fl">45</span>,</span>
<span>     breaks <span class="op">=</span> <span class="va">breaks</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="comparative-analysis_files/figure-html/dr_histo-1.png" width="672"></p>
<p>It seems like for the most part, the Tetraodontiformes have a low
speciation rate, except for a subset of non-reef species that have a
faster rate.</p>
<p>Of course, with any comparative method, it’s critical to consider the
historical relationships between the species you’re examining. The
following snippet of code is quite complex, but demonstrates how to draw
rates onto a phylogeny using colored bars next to each tip in
question.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot tree and extract plotting data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tree</span>, show.tip.label <span class="op">=</span> <span class="cn">FALSE</span>, no.margin <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html" class="external-link">get</a></span><span class="op">(</span><span class="st">"last_plot.phylo"</span>, <span class="va">.PlotPhyloEnv</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate a color ramp</span></span>
<span><span class="va">ramp</span> <span class="op">&lt;-</span> <span class="fu">grDevices</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRamp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>, bias <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">tiporder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">rates</span><span class="op">$</span><span class="va">tip</span>, <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span></span>
<span><span class="va">scaled_rates</span> <span class="op">&lt;-</span> <span class="va">rates</span><span class="op">$</span><span class="va">dr</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">rates</span><span class="op">$</span><span class="va">dr</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">tipcols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu">ramp</span><span class="op">(</span><span class="va">scaled_rates</span><span class="op">)</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rgb</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">x</span> <span class="op">/</span> <span class="fl">255</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Place colored bars</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">ii</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">tiporder</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tip</span> <span class="op">&lt;-</span> <span class="va">tiporder</span><span class="op">[</span><span class="va">ii</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">obj</span><span class="op">$</span><span class="va">xx</span><span class="op">[</span><span class="va">tip</span><span class="op">]</span> <span class="op">+</span> <span class="fl">0.5</span>, <span class="va">obj</span><span class="op">$</span><span class="va">xx</span><span class="op">[</span><span class="va">tip</span><span class="op">]</span> <span class="op">*</span> <span class="fl">1.5</span> <span class="op">+</span> <span class="fl">0.5</span> <span class="op">+</span> <span class="va">scaled_rates</span><span class="op">[</span><span class="va">ii</span><span class="op">]</span><span class="op">)</span>,</span>
<span>          y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">obj</span><span class="op">$</span><span class="va">yy</span><span class="op">[</span><span class="va">tip</span><span class="op">]</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>          col <span class="op">=</span> <span class="va">tipcols</span><span class="op">[</span><span class="va">ii</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="comparative-analysis_files/figure-html/plot_tree_dr-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="running-an-analysis">Running an analysis<a class="anchor" aria-label="anchor" href="#running-an-analysis"></a>
</h2>
<p>Let’s perform a more quantitative analysis using using
<code>hisse</code>. We’ll test 4 models: a BiSSE-like model, a
BiSSE-like null model, a hisse model, and the hisse 2 state null
model.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">hisse</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: deSolve</span></span>
<span><span class="co">#&gt; Loading required package: GenSA</span></span>
<span><span class="co">#&gt; Loading required package: subplex</span></span>
<span><span class="co">#&gt; Loading required package: nloptr</span></span></code></pre></div>
<p>The <code>hisse</code> package parameterizes things differently from
<code>diversitree</code> (where BiSSE lives), so we aren’t able to
exactly replicate the analyses in the Santini paper. Instead we’ll
settle by ensuring that the epsilon parameter, <span class="math inline">\(\epsilon = \frac{\mu}{\lambda}\)</span> is
constrained to be equal for both reef and non-reef taxa. We’ll also
constrain transition rates to be equal, since it can be difficult to
estimate those.</p>
<p>Note that to ensure this vignette can be run in a reasonable amount
of time, we set <code>sann = FALSE</code> to disable the simulated
annealing procedure in <code>hisse</code>. However, for any actual
analysis this option should be turned on for maximum accuracy and
confidence in your final results.</p>
<p>First, we’ll construct and run the BiSSE model and the BiSSE null
model:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trans.rates.bisse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/hisse/man/transMatMakerHiSSE.html" class="external-link">TransMatMakerHiSSE</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pp.bisse.full</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/hisse/man/hisse.html" class="external-link">hisse</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">reef</span>,</span>
<span>                       hidden.states <span class="op">=</span> <span class="cn">FALSE</span>, sann <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       turnover <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, eps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                       trans.rate <span class="op">=</span> <span class="va">trans.rates.bisse</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in hisse(tree, reef, hidden.states = FALSE, sann = FALSE, turnover =</span></span>
<span><span class="co">#&gt; c(1, : You have chosen to rely on the internal starting points that generally</span></span>
<span><span class="co">#&gt; work but does not guarantee finding the MLE.</span></span>
<span><span class="co">#&gt; Initializing... </span></span>
<span><span class="co">#&gt; Finished. Beginning bounded subplex routine... </span></span>
<span><span class="co">#&gt; Finished. Summarizing results...</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pp.bisse.null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/hisse/man/hisse.html" class="external-link">hisse</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">reef</span>,</span>
<span>                       hidden.states <span class="op">=</span> <span class="cn">FALSE</span>, sann <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       turnover <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, eps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                       trans.rate <span class="op">=</span> <span class="va">trans.rates.bisse</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in hisse(tree, reef, hidden.states = FALSE, sann = FALSE, turnover =</span></span>
<span><span class="co">#&gt; c(1, : You have chosen to rely on the internal starting points that generally</span></span>
<span><span class="co">#&gt; work but does not guarantee finding the MLE.</span></span>
<span><span class="co">#&gt; Initializing... </span></span>
<span><span class="co">#&gt; Finished. Beginning bounded subplex routine... </span></span>
<span><span class="co">#&gt; Finished. Summarizing results...</span></span></code></pre></div>
<p>Next, we’ll run the full hisse model, save for the constrained
transition rates and epsilon.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trans.rates.hisse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/hisse/man/transMatMakerHiSSE.html" class="external-link">TransMatMakerHiSSE</a></span><span class="op">(</span>hidden.traits <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">trans.rates.hisse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/hisse/man/transMat.html" class="external-link">ParEqual</a></span><span class="op">(</span><span class="va">trans.rates.hisse</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pp.hisse.full</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/hisse/man/hisse.html" class="external-link">hisse</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">reef</span>,</span>
<span>                       hidden.states <span class="op">=</span> <span class="cn">TRUE</span>, sann <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       turnover <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span>, eps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                       trans.rate <span class="op">=</span> <span class="va">trans.rates.hisse</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in hisse(tree, reef, hidden.states = TRUE, sann = FALSE, turnover =</span></span>
<span><span class="co">#&gt; c(1, : You have chosen to rely on the internal starting points that generally</span></span>
<span><span class="co">#&gt; work but does not guarantee finding the MLE.</span></span>
<span><span class="co">#&gt; Initializing... </span></span>
<span><span class="co">#&gt; Finished. Beginning bounded subplex routine... </span></span>
<span><span class="co">#&gt; Finished. Summarizing results...</span></span></code></pre></div>
<p>Finally, we’ll build the 2 state character independent
diversification model, sometimes called CID-2. We’ll use this as our
null model by forcing the visible states (reef or non-reef) to have the
same net turnover rates, while permitting the hidden states to vary
freely.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pp.hisse.null2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/hisse/man/hisse.html" class="external-link">hisse</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">reef</span>,</span>
<span>                        hidden.states <span class="op">=</span> <span class="cn">TRUE</span>, sann <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        turnover <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, eps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                        trans.rate <span class="op">=</span> <span class="va">trans.rates.hisse</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in hisse(tree, reef, hidden.states = TRUE, sann = FALSE, turnover =</span></span>
<span><span class="co">#&gt; c(1, : You have chosen to rely on the internal starting points that generally</span></span>
<span><span class="co">#&gt; work but does not guarantee finding the MLE.</span></span>
<span><span class="co">#&gt; Initializing... </span></span>
<span><span class="co">#&gt; Finished. Beginning bounded subplex routine... </span></span>
<span><span class="co">#&gt; Finished. Summarizing results...</span></span></code></pre></div>
<p>We can combine all of our results into a single table for easy
comparison.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">pp.bisse.full</span>, <span class="va">pp.bisse.null</span>, <span class="va">pp.hisse.null2</span>, <span class="va">pp.hisse.full</span><span class="op">)</span></span>
<span><span class="va">aicc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">results</span>, <span class="va">`[[`</span>, <span class="st">"AICc"</span><span class="op">)</span></span>
<span><span class="va">lnl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">results</span>, <span class="va">`[[`</span>, <span class="st">"loglik"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bisse_full"</span>, <span class="st">"bisse_null"</span>, <span class="st">"hisse_cid2"</span>, <span class="st">"hisse_full"</span><span class="op">)</span>, <span class="va">aicc</span>, <span class="va">lnl</span><span class="op">)</span></span>
<span><span class="co">#&gt;        model     aicc       lnl</span></span>
<span><span class="co">#&gt; 1 bisse_full 1829.695 -909.7075</span></span>
<span><span class="co">#&gt; 2 bisse_null 1827.743 -909.7785</span></span>
<span><span class="co">#&gt; 3 hisse_cid2 1816.876 -904.3447</span></span>
<span><span class="co">#&gt; 4 hisse_full 1820.233 -903.9193</span></span></code></pre></div>
<p>Summarizing the results on the basis of AICc suggests that the best
supported model is a null model, where habitat has no effect on
speciation rate.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jonathan Chang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
